pkg/tenantctl/controller.go

```diff
@@ -157,219 +172,470 @@ func (c *Controller) processNextWorkItem() bool {
+		return c.testing.updateStatusStub(tenant)
+	}
+
 	client := c.factory.StewardV1alpha1().Tenants(tenant.GetNamespace())
 	updatedTenant, err := client.UpdateStatus(tenant)
 	if err != nil {
-		err = errors.WithMessagef(err, "Failed to update status of tenant '%s' in namespace '%s'", tenant.GetName(), tenant.GetNamespace())
-		log.Printf("ERROR: %s", err.Error())
+		err = errors.WithMessage(err, "failed to update resource status")
+		c.logPrintln(tenant, err)
 		return nil, err
 	}
 	return updatedTenant, nil
 }
 
 func (c *Controller) update(tenant *api.Tenant) (*api.Tenant, error) {
 	client := c.factory.StewardV1alpha1().Tenants(tenant.GetNamespace())
-	updatedTenant, err := client.Update(tenant)
+	result, err := client.Update(tenant)
 	if err != nil {
-		err = errors.WithMessagef(err, "Failed to update tenant '%s' in namespace '%s'", tenant.GetName(), tenant.GetNamespace())
-		log.Printf("ERROR: %s", err.Error())
-		return nil, err
+		err = errors.WithMessagef(err,
+			"failed to update tenant %q in namespace %q",
+			tenant.GetName(), tenant.GetNamespace(),
+		)
+		c.logPrintln(tenant, err)
+		return tenant, err
 	}
-	return updatedTenant, nil
+	return result, nil
 }
 
-// An error is returned in cases to signalize processNextWorkItem() to retry processing the tenant.
-// If no error is returned this signalized OK, do not retry. This should be done in cases where retry will not help.
-func (c *Controller) handleError(tenant *api.Tenant, err error, result api.TenantResult) error {
-	log.Printf("ERROR: %s", err.Error())
-	tenant.Status.Result = result
-	tenant.Status.Message = utils.Trim(err.Error())
-	_, updateStatusErr := c.updateStatus(tenant)
-	return updateStatusErr
+func (c *Controller) checkNamespaceExists(name string) (bool, error) {
+	namespaces := c.factory.CoreV1().Namespaces()
+	namespace, err := namespaces.Get(name, metav1.GetOptions{})
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return false, nil
+		}
+		err = errors.WithMessagef(err, "error: failed to get namespace %q", name)
+		return false, err
+	}
+	return namespace.GetDeletionTimestamp().IsZero(), nil
 }
 
-func (c *Controller) rollbackIfRequired(tenantKey string) {
-	tenant, err := c.fetcher.ByKey(tenantKey)
+func (c *Controller) createTenantNamespace(config clientConfig, tenant *api.Tenant) (string, error) {
+	c.logPrintln(tenant, "creating new tenant namespace")
+	namespaceManager := c.getNamespaceManager(config)
+	nsName, err := namespaceManager.Create(tenant.GetName(), nil)
 	if err != nil {
-		log.Printf("ERROR: Could not get tenant during rollback: %s", err.Error())
+		err = errors.WithMessage(err, "failed to create new tenant namespace")
+		c.logPrintln(tenant, err)
+		return "", err
 	}
-	if tenant.Status.Progress != api.TenantProgressFinished {
-		_ = c.rollback(tenant)
+	return nsName, err
+}
+
+func (c *Controller) rollbackTenantNamespace(namespace string, tenant *api.Tenant, config clientConfig) error {
```

Comment:
Not sure if `rollback` is the right term here. I would use `rollback` if an operation failed and I want to recover the previous state. What about `dismantle`, `dispose`, `eliminate`?

Reply:
Fixed in ee6e8713c6ced1140aa42f99b9d8bf6106e3bafa.

I've used "delete", as this is the name of the underlying operation of the K8s API.